src/core/lib/security/authorization/audit_logging.cc

```diff
@@ -39,62 +40,65 @@
 namespace grpc_core {
 namespace experimental {
 
-void AuditLoggerRegistry::RegisterAuditLoggerFactory(
+namespace {
+
+Mutex* g_mu = new Mutex();
+AuditLoggerRegistry* g_registry ABSL_GUARDED_BY(g_mu) =
+    new AuditLoggerRegistry();
+
+}  // namespace
+
+void AuditLoggerRegistry::RegisterFactory(
     std::unique_ptr<AuditLoggerFactory> factory) {
-  auto& registry = GetAuditLoggerRegistry();
-  MutexLock lock(&registry.mu_);
-  registry.logger_factories_map_[factory->name()] = std::move(factory);
+  if (factory == nullptr) return;
+  MutexLock lock(g_mu);
+  GPR_ASSERT(g_registry->logger_factories_map_.find(factory->name()) ==
+             g_registry->logger_factories_map_.end());
+  g_registry->logger_factories_map_[factory->name()] = std::move(factory);
 }
 
-bool AuditLoggerRegistry::AuditLoggerFactoryExists(absl::string_view name) {
-  auto& registry = AuditLoggerRegistry::GetAuditLoggerRegistry();
-  MutexLock lock(&registry.mu_);
-  return registry.logger_factories_map_.find(name) !=
-         registry.logger_factories_map_.end();
+bool AuditLoggerRegistry::FactoryExists(absl::string_view name) {
+  MutexLock lock(g_mu);
+  return g_registry->logger_factories_map_.find(name) !=
+         g_registry->logger_factories_map_.end();
 }
 
 absl::StatusOr<std::unique_ptr<AuditLoggerFactory::Config>>
-AuditLoggerRegistry::ParseAuditLoggerConfig(absl::string_view name,
-                                            const Json& json) {
-  auto factory_or = AuditLoggerRegistry::GetAuditLoggerFactory(name);
+AuditLoggerRegistry::ParseConfig(absl::string_view name, const Json& json) {
+  MutexLock lock(g_mu);
+  auto factory_or = g_registry->GetAuditLoggerFactory(name);
   if (!factory_or.ok()) {
-    return absl::InvalidArgumentError("unsupported audit logger type");
+    return factory_or.status();
   }
   return factory_or.value()->ParseAuditLoggerConfig(json);
 }
 
 std::unique_ptr<AuditLogger> AuditLoggerRegistry::CreateAuditLogger(
     std::unique_ptr<AuditLoggerFactory::Config> config) {
-  auto factory_or = AuditLoggerRegistry::GetAuditLoggerFactory(config->name());
+  MutexLock lock(g_mu);
+  auto factory_or = g_registry->GetAuditLoggerFactory(config->name());
   GPR_ASSERT(factory_or.ok());
   return factory_or.value()->CreateAuditLogger(std::move(config));
 }
 
-AuditLoggerRegistry& AuditLoggerRegistry::GetAuditLoggerRegistry() {
-  static AuditLoggerRegistry& registry = *new AuditLoggerRegistry();
-  return registry;
-}
-
 absl::StatusOr<AuditLoggerFactory*> AuditLoggerRegistry::GetAuditLoggerFactory(
     absl::string_view name) {
```

Comment:
Suggest adding an `ABSL_EXCLUSIVE_LOCKS_REQUIRED(g_mu)` annotation here.

Reply:
I removed this function. There are only two places that use it and given the if condition there, putting this logic inline doesn't lead to more LOC. It also eliminates the problem of locking both registry reads and map I/O.